File: Python/gamex/platform.py
import os
from typing import Callable
from enum import Enum
from .pak import PakFile
from openstk.gfx import IObjectManager, IMaterialManager, IShaderManager, ITextureManager, PlatformStats
from openstk.gfx_render import IMaterial
from openstk.gfx_texture import ITexture

# typedefs
class Shader: pass
class Texture: pass
class Material: pass

# TextureBuilderBase
class TextureBuilderBase:
    maxTextureMaxAnisotropy: int = PlatformStats.maxTextureMaxAnisotropy
    defaultTexture: Texture
    def buildTexture(info: ITexture, rng: range = None) -> Texture: pass
    def buildSolidTexture(width: int, height: int, rgba: list[float]) -> Texture: pass
    def buildNormalMap(source: Texture, strength: float): pass
    def deleteTexture(texture: Texture): pass

# TextureManager
class TextureManager(ITextureManager):
    _pakFile: PakFile
    _builder: TextureBuilderBase
    _cachedTextures: dict[object, (Texture, dict[str, object])] = {}
    _preloadTasks: dict[object, ITexture] = {}

    def __init__(self, pakFile: PakFile, builder: TextureBuilderBase):
        self._pakFile = pakFile
        self._builder = builder

    def buildSolidTexture(self, width: int, height: int, rgba: list[float] = None) -> Texture: return self._builder.buildSolidTexture(width, height, rgba)

    def buildNormalMap(self, source: Texture, strength: float) -> Texture: return self._builder.buildNormalMap(source, strength)

    @property
    def defaultTexture(self) -> Texture: return self._builder.defaultTexture

    def loadTexture(self, key: object, rng: range = None) -> (Texture, dict[str, object]):
        if key in self._cachedTextures: return self._cachedTextures[key]
        # Load & cache the texture.
        info = key if isinstance(key, ITexture) else self.loadTexture(key)
        texture = self._builder.buildTexture(info, rng) if info else self._builder.defaultTexture
        data = info.data if info else None
        self._cachedTextures[key] = (texture, data)
        return (texture, data)

    def preloadTexture(self, path: str) -> None:
        if path in self._cachedTextures: return
        # Start loading the texture file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(path)

    def deleteTexture(self, key: object) -> None:
        if not key in self._cachedTextures: return
        self._builder.deleteTexture(self._cachedTextures[0])
        self._cachedTextures.remove(key)

    async def _loadTexture(self, key: object) -> ITexture:
        assert(not key in self._cachedTextures)
        match key:
            case s if isinstance(key, str):
                self.preloadTexture(s)
                info = await self._preloadTasks[key]
                self._preloadTasks.remove(key)
                return info
            case _: raise Exception(f'Unknown {key}')

# ShaderBuilderBase
class ShaderBuilderBase:
    def buildShader(self, path: str, args: dict[str, bool]) -> Shader: pass
    def buildPlaneShader(self, path: str, args: dict[str, bool]) -> Shader: pass

# ShaderManager
class ShaderManager(IShaderManager):
    emptyArgs: dict[str, bool] = {}

    _pakFile: PakFile
    _builder: ShaderBuilderBase

    def __init__(self, pakFile: PakFile, builder: ShaderBuilderBase):
        self._pakFile = pakFile
        self._builder = builder
    
    def loadShader(self, path: str, args: dict[str, bool] = None) -> Shader:
        return self._builder.buildShader(path, args or self.emptyArgs)

    def loadPlaneShader(self, path: str, args: dict[str, bool] = None) -> Shader:
        return self._builder.buildPlaneShader(path, args or self.emptyArgs)

# ObjectBuilderBase
class ObjectBuilderBase:
    def createObject(prefab: object) -> object: pass
    def ensurePrefabContainerExists() -> None: pass
    def buildObject(source: object, materialManager: IMaterialManager) -> object: pass

# ObjectManager
class ObjectManager(IObjectManager):
    _pakFile: PakFile
    _materialManager: IMaterialManager
    _builder: ObjectBuilderBase
    _cachedPrefabs: dict[str, object] = {}
    _preloadTasks: dict[str, object] = {}

    def __init__(self, pakFile: PakFile, materialManager: IMaterialManager, builder: ObjectBuilderBase):
        self._pakFile = pakFile
        self._materialManager = materialManager
        self._builder = builder

    def createObject(self, path: str) -> (object, dict[str, object]):
        data = None
        self._builder.ensurePrefabContainerExists()
        # Load & cache the NIF prefab.
        if not path in self._cachedPrefabs: prefab = self._cachedPrefabs[path] = self.loadPrefabDontAddToPrefabCache(path)
        else: prefab = self._cachedPrefabs[path]
        # Instantiate the prefab.
        return self._builder.createObject(prefab)
 
    def preloadObject(self, path: str) -> None:
        if path in self._cachedPrefabs: return
        # Start loading the object asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(object, path)

    async def loadPrefabDontAddToPrefabCache(path: str) -> object:
        assert(not path in self._cachedPrefabs)
        self.preloadObject(path)
        source = await self._preloadTasks[path]
        self._preloadTasks.remove(path)
        return self._builder.buildObject(source, self._materialManager)

# MaterialBuilderBase
class MaterialBuilderBase:
    textureManager : ITextureManager
    normalGeneratorIntensity: float = 0.75
    defaultMaterial: Material

    def __init__(self, textureManager: ITextureManager): self.TextureManager = textureManager
    def buildMaterial(self, key: object) -> Material: pass

# MaterialManager
class MaterialManager(IMaterialManager):
    _pakFile: PakFile
    _builder: MaterialBuilderBase
    _cachedMaterials: dict[object, (Material, dict[str, object])] = {}
    _preloadTasks: dict[object, IMaterial] = {}

    textureManager: ITextureManager

    def __init__(self, pakFile: PakFile, textureManager: ITextureManager, builder: MaterialBuilderBase):
        self._pakFile = pakFile
        self._textureManager = textureManager
        self._builder = builder

    def loadMaterial(self, key: object) -> (Material, dict[str, object]):
        if key in self._cachedMaterials: return self._cachedMaterials[key]
        # Load & cache the material.
        info = key if isinstance(key, IMaterial) else self.loadMaterialInfo(key)
        material = self._builder.buildMaterial(info) if info else self._builder.defaultMaterial
        data = info.data if info else None
        self._cachedMaterials[key] = (material, data)
        return material

    def preloadMaterial(self, path: str) -> None:
        if path in self._cachedMaterials: return
        # Start loading the material file asynchronously if we haven't already started.
        if not path in self._preloadTasks: self._preloadTasks[path] = self._pakFile.loadFileObject(path)

    async def loadMaterialInfo(self, key: object) -> IMaterial:
        assert(not key in self._cachedMaterials)
        match key:
            case path if isinstance(key, str):
                self.preloadMaterial(path)
                info = await self.preloadTasks[key]
                self.preloadTasks.remove(key)
                return info
            case _: raise Exception(f'Unknown {key}')

# typedefs
# class Type(Enum): pass
# class OS(Enum): pass

# Platform
class Platform:
    class Type(Enum):
        Unknown = 0
        OpenGL = 1
        Unity = 2
        Vulken = 3
        Test = 4
        Other = 5

    class OS(Enum):
        Windows = 1
        OSX = 1
        Linux = 1
        Android = 1

    platformType: Type = None
    platformTag: str = None
    platformOS: OS = OS.Windows
    graphicFactory: Callable = None
    startups: list[object] = []
    inTestHost: bool = False
    logFunc: Callable = None

    class Stats:
        maxTextureMaxAnisotropy: int

    # startup
    @staticmethod
    def startup() -> None:
        if Platform.inTestHost and len(Platform.startups) == 0: Platform.startups.append(TestPlatform.startup)
        for startup in Platform.startups:
            if startup(): return
        Platform.platformType = Platform.Type.Unknown
        Platform.graphicFactory = lambda source: None
        Platform.logFunc = lambda a: print(a)

# TestGraphic
class TestGraphic:
    def __init__(self, source):
        self._source = source

# TestPlatform
class TestPlatform:
    def startup() -> bool:
        Platform.platformType = Platform.Type.Test
        Platform.graphicFactory = lambda source: TestGraphic(source)
        Platform.logFunc = lambda a: print(a)
        return True

File: dotnet/Base/GameX.Base/Platform.cs
using OpenStack;
using OpenStack.Graphics;
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.InteropServices;
using System.Threading.Tasks;
using static OpenStack.Debug;
using Stopwatch = System.Diagnostics.Stopwatch;

namespace GameX.Platforms
{
    #region TextureBuilderBase

    public abstract class TextureBuilderBase<Texture>
    {
        public static int MaxTextureMaxAnisotropy
        {
            get => PlatformStats.MaxTextureMaxAnisotropy;
            set => PlatformStats.MaxTextureMaxAnisotropy = value;
        }

        public abstract Texture DefaultTexture { get; }
        public abstract Texture BuildTexture(ITexture info, Range? rng = null);
        public abstract Texture BuildSolidTexture(int width, int height, float[] rgba);
        public abstract Texture BuildNormalMap(Texture source, float strength);
        public abstract void DeleteTexture(Texture texture);
    }

    #endregion

    #region TextureManager

    public class TextureManager<Texture> : ITextureManager<Texture>
    {
        readonly PakFile PakFile;
        readonly TextureBuilderBase<Texture> Builder;
        readonly Dictionary<object, (Texture texture, IDictionary<string, object> data)> CachedTextures = new Dictionary<object, (Texture texture, IDictionary<string, object> data)>();
        readonly Dictionary<object, Task<ITexture>> PreloadTasks = new Dictionary<object, Task<ITexture>>();

        public TextureManager(PakFile pakFile, TextureBuilderBase<Texture> builder)
        {
            PakFile = pakFile;
            Builder = builder;
        }

        public Texture BuildSolidTexture(int width, int height, params float[] rgba) => Builder.BuildSolidTexture(width, height, rgba);

        public Texture BuildNormalMap(Texture source, float strength) => Builder.BuildNormalMap(source, strength);

        public Texture DefaultTexture => Builder.DefaultTexture;

        public Texture LoadTexture(object key, out IDictionary<string, object> data, Range? rng = null)
        {
            if (CachedTextures.TryGetValue(key, out var cache)) { data = cache.data; return cache.texture; }
            // Load & cache the texture.
            var info = key is ITexture z ? z : LoadTexture(key);
            var texture = info != null ? Builder.BuildTexture(info, rng) : Builder.DefaultTexture;
            data = info?.Data;
            CachedTextures[key] = (texture, data);
            return texture;
        }

        public void PreloadTexture(string path)
        {
            if (CachedTextures.ContainsKey(path)) return;
            // Start loading the texture file asynchronously if we haven't already started.
            if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<ITexture>(path);
        }

        public void DeleteTexture(object key)
        {
            if (!CachedTextures.TryGetValue(key, out var cache)) return;
            Builder.DeleteTexture(cache.texture);
            CachedTextures.Remove(key);
        }

        ITexture LoadTexture(object key)
        {
            Assert(!CachedTextures.ContainsKey(key));
            switch (key)
            {
                case string path:
                    PreloadTexture(path);
                    var info = PreloadTasks[key].Result;
                    PreloadTasks.Remove(key);
                    return info;
                default: throw new ArgumentOutOfRangeException(nameof(key), $"{key}");
            }
        }
    }

    #endregion

    #region ShaderBuilderBase

    public abstract class ShaderBuilderBase<Shader>
    {
        public abstract Shader BuildShader(string path, IDictionary<string, bool> args);
        public abstract Shader BuildPlaneShader(string path, IDictionary<string, bool> args);
    }

    #endregion

    #region ShaderManager

    public class ShaderManager<Shader> : IShaderManager<Shader>
    {
        static readonly Dictionary<string, bool> EmptyArgs = new Dictionary<string, bool>();
        readonly PakFile _pakFile;
        readonly ShaderBuilderBase<Shader> _builder;

        public ShaderManager(PakFile pakFile, ShaderBuilderBase<Shader> builder)
        {
            _pakFile = pakFile;
            _builder = builder;
        }

        public Shader LoadShader(string path, IDictionary<string, bool> args = null)
            => _builder.BuildShader(path, args ?? EmptyArgs);

        public Shader LoadPlaneShader(string path, IDictionary<string, bool> args = null)
            => _builder.BuildPlaneShader(path, args ?? EmptyArgs);
    }

    #endregion

    #region ObjectBuilderBase

    public abstract class ObjectBuilderBase<Object, Material, Texture>
    {
        public abstract Object CreateObject(Object prefab);
        public abstract void EnsurePrefabContainerExists();
        public abstract Object BuildObject(object source, IMaterialManager<Material, Texture> materialManager);
    }

    #endregion

    #region ObjectManager

    public class ObjectManager<Object, Material, Texture> : IObjectManager<Object, Material, Texture>
    {
        readonly PakFile _pakFile;
        readonly IMaterialManager<Material, Texture> _materialManager;
        readonly ObjectBuilderBase<Object, Material, Texture> _builder;
        readonly Dictionary<string, Object> _cachedPrefabs = new Dictionary<string, Object>();
        readonly Dictionary<string, Task<object>> _preloadTasks = new Dictionary<string, Task<object>>();

        public ObjectManager(PakFile pakFile, IMaterialManager<Material, Texture> materialManager, ObjectBuilderBase<Object, Material, Texture> builder)
        {
            _pakFile = pakFile;
            _materialManager = materialManager;
            _builder = builder;
        }

        public Object CreateObject(string path, out IDictionary<string, object> data)
        {
            data = null;
            _builder.EnsurePrefabContainerExists();
            // Load & cache the NIF prefab.
            if (!_cachedPrefabs.TryGetValue(path, out var prefab)) prefab = _cachedPrefabs[path] = LoadPrefabDontAddToPrefabCache(path);
            // Instantiate the prefab.
            return _builder.CreateObject(prefab);
        }

        public void PreloadObject(string path)
        {
            if (_cachedPrefabs.ContainsKey(path)) return;
            // Start loading the object asynchronously if we haven't already started.
            if (!_preloadTasks.ContainsKey(path)) _preloadTasks[path] = _pakFile.LoadFileObject<object>(path);
        }

        Object LoadPrefabDontAddToPrefabCache(string path)
        {
            Assert(!_cachedPrefabs.ContainsKey(path));
            PreloadObject(path);
            var source = _preloadTasks[path].Result;
            _preloadTasks.Remove(path);
            return _builder.BuildObject(source, _materialManager);
        }
    }

    #endregion

    #region MaterialBuilderBase

    public abstract class MaterialBuilderBase<Material, Texture>
    {
        protected ITextureManager<Texture> TextureManager;

        public MaterialBuilderBase(ITextureManager<Texture> textureManager) => TextureManager = textureManager;

        public float? NormalGeneratorIntensity = 0.75f;
        public abstract Material DefaultMaterial { get; }
        public abstract Material BuildMaterial(object key);
    }

    #endregion

    #region MaterialManager

    /// <summary>
    /// Manages loading and instantiation of materials.
    /// </summary>
    public class MaterialManager<Material, Texture> : IMaterialManager<Material, Texture>
    {
        readonly PakFile PakFile;
        readonly MaterialBuilderBase<Material, Texture> Builder;
        readonly Dictionary<object, (Material material, IDictionary<string, object> data)> CachedMaterials = new Dictionary<object, (Material material, IDictionary<string, object> data)>();
        readonly Dictionary<object, Task<IMaterial>> PreloadTasks = new Dictionary<object, Task<IMaterial>>();

        public ITextureManager<Texture> TextureManager { get; }

        public MaterialManager(PakFile pakFile, ITextureManager<Texture> textureManager, MaterialBuilderBase<Material, Texture> builder)
        {
            PakFile = pakFile;
            TextureManager = textureManager;
            Builder = builder;
        }

        public Material LoadMaterial(object key, out IDictionary<string, object> data)
        {
            if (CachedMaterials.TryGetValue(key, out var cache)) { data = cache.data; return cache.material; }
            // Load & cache the material.
            var info = key is IMaterial z ? z : LoadMaterialInfo(key);
            var material = info != null ? Builder.BuildMaterial(info) : Builder.DefaultMaterial;
            data = info?.Data;
            CachedMaterials[key] = (material, data);
            return material;
        }

        public void PreloadMaterial(string path)
        {
            if (CachedMaterials.ContainsKey(path)) return;
            // Start loading the material file asynchronously if we haven't already started.
            if (!PreloadTasks.ContainsKey(path)) PreloadTasks[path] = PakFile.LoadFileObject<IMaterial>(path);
        }

        IMaterial LoadMaterialInfo(object key)
        {
            Assert(!CachedMaterials.ContainsKey(key));
            switch (key)
            {
                case string path:
                    PreloadMaterial(path);
                    var info = PreloadTasks[key].Result;
                    PreloadTasks.Remove(key);
                    return info;
                default: throw new ArgumentOutOfRangeException(nameof(key), $"{key}");
            }
        }
    }

    #endregion

    #region Platform

    /// <summary>
    /// Platform
    /// </summary>
    public static class Platform
    {
        /// <summary>
        /// The platform stats.
        /// </summary>
        public class Stats
        {
            static readonly bool _HighRes = Stopwatch.IsHighResolution;
            static readonly double _HighFrequency = 1000.0 / Stopwatch.Frequency;
            static readonly double _LowFrequency = 1000.0 / TimeSpan.TicksPerSecond;
            static bool _UseHRT = false;

            public static bool UsingHighResolutionTiming => _UseHRT && _HighRes && !Unix;
            public static long TickCount => (long)Ticks;
            public static double Ticks => _UseHRT && _HighRes && !Unix ? Stopwatch.GetTimestamp() * _HighFrequency : DateTime.UtcNow.Ticks * _LowFrequency;

            public static readonly bool Is64Bit = Environment.Is64BitProcess;
            public static bool MultiProcessor { get; private set; }
            public static int ProcessorCount { get; private set; }
            public static bool Unix { get; private set; }
            public static bool VR { get; private set; }
        }

        /// <summary>
        /// The platform type.
        /// </summary>
        public enum Type { Unknown, OpenGL, Unity, Unreal, Vulken, StereoKit, Test, Other }

        /// <summary>
        /// The platform OS.
        /// </summary>
        public enum OS { Windows, OSX, Linux, Android }

        /// <summary>
        /// Gets or sets the platform.
        /// </summary>
        public static Type PlatformType;

        /// <summary>
        /// Gets or sets the platform tag.
        /// </summary>
        public static string PlatformTag;

        /// <summary>
        /// Gets the platform os.
        /// </summary>
        public static readonly OS PlatformOS = RuntimeInformation.IsOSPlatform(OSPlatform.Windows) ? OS.Windows
            : RuntimeInformation.IsOSPlatform(OSPlatform.OSX) ? OS.OSX
            : RuntimeInformation.OSDescription.StartsWith("android-") ? OS.Android
            : RuntimeInformation.IsOSPlatform(OSPlatform.Linux) ? OS.Linux
            : throw new ArgumentOutOfRangeException(nameof(RuntimeInformation.IsOSPlatform), RuntimeInformation.OSDescription);

        /// <summary>
        /// Gets or sets the platform graphics factory.
        /// </summary>
        public static Func<PakFile, IOpenGraphic> GraphicFactory;

        /// <summary>
        /// Gets the platform startups.
        /// </summary>
        public static readonly List<Func<bool>> Startups = new List<Func<bool>>();

        /// <summary>
        /// Determines if in a test host.
        /// </summary>
        public static bool InTestHost => AppDomain.CurrentDomain.GetAssemblies().Any(x => x.FullName.StartsWith("testhost,"));
    }

    #endregion

    #region TestGraphic

    public interface ITestGraphic : IOpenGraphic { }

    public class TestGraphic : ITestGraphic
    {
        readonly PakFile _source;

        public TestGraphic(PakFile source) => _source = source;
        public object Source => _source;
        public Task<T> LoadFileObject<T>(string path) => throw new NotSupportedException();
        public void PreloadTexture(string texturePath) => throw new NotSupportedException();
        public void PreloadObject(string filePath) => throw new NotSupportedException();
    }

    #endregion

    #region TestPlatform

    public static class TestPlatform
    {
        public static bool Startup()
        {
            try
            {
                Platform.PlatformType = Platform.Type.Test;
                Platform.GraphicFactory = source => new TestGraphic(source);
                Debug.AssertFunc = x => System.Diagnostics.Debug.Assert(x);
                Debug.LogFunc = a => System.Diagnostics.Debug.Print(a);
                Debug.LogFormatFunc = (a, b) => System.Diagnostics.Debug.Print(a, b);
                return true;
            }
            catch { return false; }
        }
    }

    #endregion
}